pkg/controller/datavolume-controller.go

```diff
@@ -2722,3 +2777,292 @@ func getDeltaTTL(dv *cdiv1.DataVolume, ttl int32) time.Duration {
 	}
 	return delta
 }
+
+// detectCloneSize obtains and assigns the original PVC's size when cloning using an empty storage value
+func (r *DatavolumeReconciler) detectCloneSize(
+	dv *cdiv1.DataVolume,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	cloneType cloneStrategy) error {
+
+	var targetSize resource.Quantity
+	sourcePvc, err := r.findSourcePvc(dv)
+	if err != nil {
+		return err
+	}
+	sourceSize := sourcePvc.Status.Capacity.Storage()
+
+	// Due to possible filesystem overhead complications when cloning
+	// using host-assisted strategy, we create a pod that automatically
+	// collects the size of the original virtual image with 'qemu-img'.
+	// If another strategy is used or the original PVC's volume mode
+	// is "block", we simply extract the value from the original PVC's spec.
+	if cloneType == HostAssistedClone &&
+		getVolumeMode(sourcePvc) == corev1.PersistentVolumeFilesystem &&
+		GetContentType(sourcePvc) == string(cdiv1.DataVolumeKubeVirt) {
+		// If available, we first try to get the virtual size from previous iterations
+		imgSize, available := getSizeFromAnnotations(sourcePvc)
+		if !available {
+			imgSize, err = r.getSizeFromPod(sourcePvc, dv)
+			if err != nil {
+				return err
+			} else if imgSize == 0 {
+				return nil
```

Comment:
Do I understand correctly that when the pod starts, but did not complete its job the size is `zero` and then controller continues with the clone and fails when trying to `createPvcForDatavolume`. After that it will retry.

Reply:
Yeah, that's right. Maybe I could add a return flag again to `detectCloneSize` so, if the pod is not finished, we reconcile without error.

Reply:
if that does not make code much more complex I would like an explicit error here.

Reply:
When `imgSize` is zero? I had first an error there, but since it's really an expected behavior, @awels recommended me to just return `nil` instead. Every time that happens, we record an event to the dataVolume in `getSizeFromPod`.

However, if for any reason it's not recommended to attempt the clone with `zero` size as I currently do, I could also return a flag in `detectCloneSize` alongside the error to reconcile in `reconcileClone` until the size has been obtained.

Something like:

```go
func (r *DatavolumeReconciler) reconcileClone(

         ...

        // Done would be false if imgSize is 0
	done, err := r.detectCloneSize(datavolume, pvcSpec, selectedCloneStrategy)
		if err != nil {
			return reconcile.Result{}, err
		} else if !done {
                        return reconcile.Result{}, nil
                }
	}
````

Reply:
In that case, I think it's more clear to add a new return value to `detectCloneSize`, as shown above. I'll quickly fix that :+1: